src/core/thread/mle_router.cpp

```diff
@@ -1284,6 +1284,10 @@ otError MleRouter::HandleAdvertisement(const Message &         aMessage,
         if (processRouteTlv)
         {
             SuccessOrExit(error = ProcessRouteTlv(route));
+            if (Get<RouterTable>().GetRouterIndex(static_cast<Router &>(*aNeighbor)) < kMaxRouters)
+            {
+                aNeighbor = NULL; // aNeighbor is no longer valid after `ProcessRouteTlv`
```

Comment:
Could you please give me short explanation why aNeighbor is invalid after ProcessRouteTlv?
Im trying to apply your patch to older OT code that we use currently, I don't have aNeighbor parameter in HandleAdvertisement (added by Jonathan) and Im trying to understand the fix. Thanks

Reply:
The condition `Get<RouterTable>().GetRouterIndex(static_cast<Router &>(*aNeighbor)) < kMaxRouters` is satisfied when `aNeighbor` references to a `Router` in the router table `mRouterTable`. 

It is all about how router table organize `Routers` in `mRouters`. **Note that the `Routers` in `mRouters` are ordered by `Router ID`.**
Let's assume that the current valid Router ID set is {1, 3, 5} and `aNeighbor` points to `Router 5`, i.e. `aNeighbor == &mRouters[2]`. If the new Router ID set is {1, 5} in the RouterTLV, after `ProcessRouteTlv`, `Router 5` is moved from `mRouters[2]` to `mRouters[1]`, however the value of `aNeighbor` is not updated accordingly. So, after `ProcessRouterTLV`, `aNeighbor` is still pointing to `&mRouters[2]`, which is actually an **INVALID** `Router`. 

This fix sets `aNeighbor = NULL` in such case so that the `aNeighbor` handling code after the exit label won't be executed.

Reply:
Thanks, this wasnt obvious :)
Adding routerTableChanged result to a ProcessRouteTlv function would improve readability of this code.

Don't we have a problem here that the whole block of code for a kRoleChild:
if (aNeighbor == &mParent) { ... } (line 1301) will never execute?
From what I see processRouteTlv is true when a child receives an advertisement from its parent.

Reply:
@mwilniewiec
Why do you think `if (aNeighbor == &mParent) { ... } (line 1301)` will never execute? 
Note that when `aNeighbor == &mParent`, `aNeighbor` is not pointing to a `Router`, so it won't be set to `NULL` after `ProcessRouteTLV`.

Reply:
Ok, after looking deeper now I see.
GetNeighbor will return mParent for a child node (even though it is an FTD and has the same entry in its router table) but for a router it returns a pointer to router table entry.
In that case this should work as expected.

Reply:
Thanks for your investigation.